zerver/lib/actions.py

```diff
@@ -4845,70 +4846,178 @@ def do_change_can_create_users(user_profile: UserProfile, value: bool) -> None:
+    history_public_to_subscribers: Optional[bool] = None,
+    is_web_public: Optional[bool] = None,
+    acting_user: UserProfile,
+) -> None:
+    old_invite_only_value = stream.invite_only
+    old_history_public_to_subscribers_value = stream.history_public_to_subscribers
+    old_is_web_public_value = stream.is_web_public
+
+    with transaction.atomic():
+        if is_web_public:
+            stream.is_web_public = True
+            stream.invite_only = False
+            stream.history_public_to_subscribers = True
+        else:
+            assert invite_only is not None
+            history_public_to_subscribers = get_default_value_for_history_public_to_subscribers(
+                stream.realm,
+                invite_only,
+                history_public_to_subscribers,
+            )
+            stream.invite_only = invite_only
+            stream.history_public_to_subscribers = history_public_to_subscribers
+            stream.is_web_public = False
+
+        stream.save(update_fields=["invite_only", "history_public_to_subscribers", "is_web_public"])
+
+        if old_invite_only_value != stream.invite_only:
+            RealmAuditLog.objects.create(
+                realm=stream.realm,
+                acting_user=acting_user,
+                modified_stream=stream,
+                event_type=RealmAuditLog.STREAM_INVITE_ONLY_CHANGED,
+                event_time=timezone_now(),
+                extra_data=orjson.dumps(
+                    {
+                        RealmAuditLog.OLD_VALUE: old_invite_only_value,
+                        RealmAuditLog.NEW_VALUE: stream.invite_only,
+                    }
+                ).decode(),
+            )
+
+        if old_history_public_to_subscribers_value != stream.history_public_to_subscribers:
+            RealmAuditLog.objects.create(
+                realm=stream.realm,
+                acting_user=acting_user,
+                modified_stream=stream,
+                event_type=RealmAuditLog.STREAM_HISTORY_PUBLIC_TO_SUBSCRIBERS_CHANGED,
+                event_time=timezone_now(),
+                extra_data=orjson.dumps(
+                    {
+                        RealmAuditLog.OLD_VALUE: old_history_public_to_subscribers_value,
+                        RealmAuditLog.NEW_VALUE: stream.history_public_to_subscribers,
+                    }
+                ).decode(),
+            )
+
+        if old_is_web_public_value != stream.is_web_public:
+            RealmAuditLog.objects.create(
+                realm=stream.realm,
+                acting_user=acting_user,
+                modified_stream=stream,
+                event_type=RealmAuditLog.STREAM_IS_WEB_PUBLIC_CHANGED,
+                event_time=timezone_now(),
+                extra_data=orjson.dumps(
+                    {
+                        RealmAuditLog.OLD_VALUE: old_is_web_public_value,
+                        RealmAuditLog.NEW_VALUE: stream.is_web_public,
+                    }
```

Comment:
I think we want all of these to be a common `STREAM_PROPERTY_CHANGED` audit log type, using the field name as a `property` key, like we do `do_set_realm_property`.  

```
    event_time = timezone_now()                                                                          
    RealmAuditLog.objects.create(                                                                        
        realm=realm,                                                                                     
        event_type=RealmAuditLog.REALM_PROPERTY_CHANGED,                                                 
        event_time=event_time,                                                                           
        acting_user=acting_user,                                                                         
        extra_data=orjson.dumps(                                                                         
            {                                                                                            
                RealmAuditLog.OLD_VALUE: old_value,                                                      
                RealmAuditLog.NEW_VALUE: value,                                                          
                "property": name,                                                                        
            }                                                                                            
        ).decode(),                                                                                      
    )                                                                                                    
```